c/src/lib/libbsp/arm/shared/stm32fxxxx/spi/hal-spi.c

```diff
@@ -0,0 +1,507 @@
+   _Assert(sc == RTEMS_SUCCESSFUL);
+}
+
+/**
+ * SPI Initiate Transfer
+ */
+static int stm32_spi_transfer(
+  spi_bus *base,
+  spi_msg *msgs,
+  uint32_t msg_count
+)
+{
+   rtems_status_code sc = RTEMS_SUCCESSFUL;
+   stm32_spi_bus *bus = (stm32_spi_bus *) base;
+   rtems_interval tickstart;
+
+   _Assert(base != NULL);
+
+   spi_msg *message = msgs;
+
+   tickstart = rtems_clock_get_ticks_since_boot();
+
+   /* Wait for the SPI bus to be ready */
+   while (HAL_SPI_GetState(&bus->handle) != HAL_SPI_STATE_READY)
+   {
+      if((rtems_clock_get_ticks_since_boot() - tickstart) >  SPI_WAIT_TIMEOUT)
+         return -1;
+   }
+
+   /* Hold CS Pin Low to activate Slave */
+   HAL_GPIO_WritePin(stm32_spi_get_gpio_port(message->slave_select.port), GET_PIN(message->slave_select.pin), GPIO_PIN_RESET);
+
+   /* Onlt Write */
+   if(msgs->flags == SPI_M_WR)
+   {
+      if(HAL_SPI_Transmit_IT(&bus->handle, (uint8_t*)message->pTxBuf, message->len)!= HAL_OK){
+         return -1;
+      }
+   }
+   /* Only Read */
+   else if(msgs->flags == SPI_M_RD)
+   {
+      if(HAL_SPI_Receive_IT(&bus->handle, (uint8_t*)message->pRxBuf, message->len)!= HAL_OK){
+         return -1;
+      }
+   }
+   /* Write followed by read */
+   else if (msgs->flags == SPI_M_WR_RD)
+   {
+      if(HAL_SPI_TransmitReceive_IT(&bus->handle, (uint8_t*)message->pTxBuf, (uint8_t*)message->pRxBuf, message->len)!= HAL_OK){
+         return -1;
+      }
+   }
+
+   /* Hold CS Pin High to de-activate Slave */
+   HAL_GPIO_WritePin(stm32_spi_get_gpio_port(message->slave_select.port), GET_PIN(message->slave_select.pin), GPIO_PIN_SET);
+
+   bus->task_id = rtems_task_self();
+
+   sc = rtems_event_transient_receive(RTEMS_WAIT, 20);
+
+   if (sc != RTEMS_SUCCESSFUL) {
+      return -ETIMEDOUT;
+   }
+}
+
+/**
+ * Initalize GPIO Pins
+ * Note: If software based slave select, gpio pin for the slave select pin
+ *      has to be configured and initialized by the application.
+ */
+void stm32_spi_gpio_init(SPI_Instance spi_instance)
+{
+   GPIO_InitTypeDef  GPIO_InitStruct;
+
+   /* Initalize SPI Clock */
+   stm32_spi_initialize_spi_clock(spi_instance);
+
+   /* Initalize GPIO Clock */
+   //TODO:: Initialize port based clock. To be implemented in gpio driver
+   __HAL_RCC_GPIOA_CLK_ENABLE();
+   __HAL_RCC_GPIOB_CLK_ENABLE();
+   __HAL_RCC_GPIOC_CLK_ENABLE();
+   __HAL_RCC_GPIOD_CLK_ENABLE();
+   __HAL_RCC_GPIOE_CLK_ENABLE();
+
+  GPIO_InitStruct.Mode      = GPIO_MODE_AF_PP;
+  GPIO_InitStruct.Pull      = GPIO_PULLDOWN;
+  GPIO_InitStruct.Speed     = GPIO_SPEED_FAST;
+  GPIO_InitStruct.Alternate = pin_config[spi_instance].alternate_func;
+
+  /* SPI SCK Pin */
+  GPIO_InitStruct.Pin       = pin_config[spi_instance].sck_pin;
+  HAL_GPIO_Init(pin_config[spi_instance].port, &GPIO_InitStruct);
+
+  /* SPI MOSI Pin */
+  GPIO_InitStruct.Pin       = pin_config[spi_instance].mosi_pin;
+  HAL_GPIO_Init(pin_config[spi_instance].port, &GPIO_InitStruct);
+
```

Comment:
The SS pin should be congifured by the application. This probably gives the flexibility to the user application to define which slave(s) should be connected to which pin(s).